engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+
+            final WritableRowRedirection rowRedirection = stateManager.getRowRedirectionForTable(tableNumber);
+            if (refreshing) {
+                rowRedirection.startTrackingPrevValues();
+            }
+            final Table inputTable = joinInput.inputTable();
+
+            for (final JoinAddition ja : joinInput.columnsToAdd()) {
+                resultSources.put(ja.newColumn().name(), RedirectedColumnSource.alwaysRedirect(rowRedirection,
+                        inputTable.getColumnSource(ja.existingColumn().name())));
+            }
+        }
+
+        final QueryTable result =
+                new QueryTable(RowSetFactory.flat(stateManager.getResultSize()).toTracking(), resultSources);
+
+        if (refreshing) {
+            final ModifiedColumnSet[] resultModifiedColumnSet = new ModifiedColumnSet[useJoinInputs.length];
+            final List<MultiJoinListenerRecorder> listenerRecorders = Collections.synchronizedList(new ArrayList<>());
+
+            final Logger log = LoggerFactory.getLogger(MultiJoinTableImpl.class);
+
+            final MergedListener mergedListener = new MultiJoinMergedListener(
+                    (IncrementalMultiJoinStateManagerTypedBase) stateManager,
+                    listenerRecorders,
+                    Collections.emptyList(),
+                    "multiJoin(" + keyColumns + ")",
+                    result,
+                    resultModifiedColumnSet);
+
+            for (int ii = 0; ii < useJoinInputs.length; ++ii) {
+                final MultiJoinInput jd = useJoinInputs[ii];
+                if (jd.inputTable().isRefreshing()) {
+                    final QueryTable input = (QueryTable) jd.inputTable();
+                    final ColumnSource<?>[] keySources = getKeySources(jd);
+
+
+                    final String[] matchNames =
+                            Arrays.stream(jd.columnsToMatch()).map(v -> v.right().name()).toArray(String[]::new);
+                    final ModifiedColumnSet sourceKeyModifiedColumnSet =
+                            input.newModifiedColumnSet(matchNames);
+
+                    final String[] addNames =
+                            Arrays.stream(jd.columnsToAdd()).map(v -> v.newColumn().name()).toArray(String[]::new);
+                    resultModifiedColumnSet[ii] =
+                            result.newModifiedColumnSet(addNames);
+                    final ModifiedColumnSet.Transformer transformer =
+                            input.newModifiedColumnSetTransformer(result, addNames);
+
+                    final MultiJoinListenerRecorder listenerRecorder =
+                            new MultiJoinListenerRecorder("multiJoin(" + ii + ")", input, result, keySources,
+                                    sourceKeyModifiedColumnSet, transformer, ii);
+                    listenerRecorder.setMergedListener(mergedListener);
+                    input.addUpdateListener(listenerRecorder);
+                    listenerRecorders.add(listenerRecorder);
```

Comment:
1. The listener recorder does not need to be added as a parent reference for dependent. That relationship is typically enforced via the merged listener.
2. The merged listener should manage the listener recorders. Since it doesn't see them at construction time, you need to make that explicit here.

The best way to accomplish both is to pass the merged listener as dependent to listener recorder construction, instead of passing the result.